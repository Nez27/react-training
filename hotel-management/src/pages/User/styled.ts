import styled from 'styled-components';

const StyledUser = styled.main`
  padding: 20px;
  padding-bottom: 100px;

  display: flex;
  flex-direction: column;
  gap: 30px;
`;

const Title = styled.h2`
  font-size: var(--fs-md);
  color: var(--dark-text);
  text-transform: capitalize;
`;

const StyledOperationTable = styled.div`
  display: flex;
  gap: 30px;
  justify-content: flex-end;
`;

export {
  StyledUser,
  Title,
  StyledOperationTable
};
