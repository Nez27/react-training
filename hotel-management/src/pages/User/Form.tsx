import { useEffect, useState } from 'react';
import toast from 'react-hot-toast';
import { FormProvider, useForm } from 'react-hook-form';

// Styled
import Input from '../../commons/styles/Input';
import TextArea from '../../commons/styles/TextArea';

// Components
import Form from '../../components/Form';
import FormRow from '../../components/LabelControl/index.tsx';

// Utils
import { sendRequest } from '../../helpers/sendRequest.ts';
import {
  isEmptyObj,
  isValidNumber,
  isValidPhoneNumber,
  isValidString,
} from '../../helpers/validators.ts';

// Constants
import { STATUS_CODE } from '../../constants/responseStatus.ts';
import {
  ADD_SUCCESS,
  EDIT_SUCCESS,
  errorMsg,
} from '../../constants/messages.ts';
import { USER_PATH } from '../../constants/path.ts';
import Select, { ISelectOptions } from '../../components/Select';

// Hooks

// Styled
import { FormBtn } from './styled.ts';
import {
  INVALID_FIELD,
  INVALID_PHONE,
  REQUIRED_FIELD_ERROR,
} from '../../constants/formValidateMessage.ts';
import { getAllRoom, updateRoomStatus } from '../../services/roomServices.ts';
import { TUser } from '../../globals/types.ts';
import { INIT_VALUE_USER_FORM } from '../../constants/variables.ts';

interface IUserFormProp {
  onClose: () => void;
  reload: boolean;
  setReload: React.Dispatch<React.SetStateAction<boolean>>;
  user?: TUser | null;
  isAdd: boolean;
}

const UserForm = ({
  onClose,
  reload,
  setReload,
  user,
  isAdd,
}: IUserFormProp) => {
  const formMethods = useForm<TUser>();
  const {
    register,
    handleSubmit,
    reset,
    formState: { errors, isDirty, isValid },
    trigger,
  } = formMethods;
  const [options, setOptions] = useState<ISelectOptions[]>();

  // Init value when edit form and load options
  useEffect(() => {
    const load = async () => {
      const rooms = await getAllRoom();
      const options: ISelectOptions[] = [];
      const tempUser = isAdd 
        ? {}
        : {...user};

      // Load and set default options room
      if (rooms) {
        rooms.forEach((item) => {
          if (!item.status || tempUser?.roomId === item.id)
            options.push({
              label: item.name! as string,
              value: item.id!.toString(),
            });
        });

        setOptions(options);
        reset({ roomId: +options[0].value });
      }

      if (!isEmptyObj(tempUser)) {
        // Init value
        // Set default value when user not have room yet.
        if (!tempUser.roomId) {
          tempUser.roomId = +options[0].value;
        }

        reset(tempUser);
      } else {
        reset(INIT_VALUE_USER_FORM);
      }
    };

    load();
  }, [reset, user, isAdd]);

  // Submit form
  const onSubmit = async (newUser: TUser) => {
    try {
      if (isAdd) {
        // Add request
        const response = await sendRequest(
          USER_PATH,
          'POST',
          JSON.stringify(newUser)
        );

        if (response.statusCode === STATUS_CODE.CREATE) {
          toast.success(ADD_SUCCESS);
        } else {
          throw new Error(errorMsg(response.statusCode, response.msg));
        }

        // Update room status
        updateRoomStatus(newUser.roomId, true);
      } else {
        // Edit request
        const response = await sendRequest(
          USER_PATH + `/${newUser.id}`,
          'PUT',
          JSON.stringify(newUser)
        );

        if (response.statusCode == STATUS_CODE.OK) {
          toast.success(EDIT_SUCCESS);
        } else {
          throw new Error(errorMsg(response.statusCode, response.msg));
        }

        // Update room status
        updateRoomStatus(user!.roomId, true, newUser.roomId);
      }
      // Reload table data
      setReload(!reload);
    } catch (error: unknown) {
      if (error instanceof Error) {
        toast.error(error.message);
      }
    }

    reset();
    onClose();
  };

  return (
    <FormProvider {...formMethods}>
      <Form onSubmit={handleSubmit(onSubmit)}>
        <Input type="hidden" id="id" {...register('id')} />
        <FormRow label="Full Name" error={errors?.name?.message}>
          <Input
            type="text"
            id="name"
            {...register('name', {
              required: REQUIRED_FIELD_ERROR,
              validate: {
                checkValidName: (value) =>
                  isValidString(value) || INVALID_FIELD,
              },
              onChange: () => trigger('name'),
            })}
          />
        </FormRow>

        <FormRow
          label="Identified Code"
          error={errors?.identifiedCode?.message}
        >
          <Input
            type="text"
            id="identifiedCode"
            {...register('identifiedCode', {
              required: REQUIRED_FIELD_ERROR,
              validate: {
                checkIdentifiedCode: (v) => isValidNumber(v) || INVALID_FIELD,
              },
              onChange: () => trigger('identifiedCode'),
            })}
          />
        </FormRow>

        <FormRow label="Phone" error={errors?.phone?.message}>
          <Input
            type="text"
            id="phone"
            {...register('phone', {
              required: REQUIRED_FIELD_ERROR,
              validate: {
                checkPhoneNum: (v) => isValidPhoneNumber(v) || INVALID_PHONE,
              },
              onChange: () => trigger('phone'),
            })}
          />
        </FormRow>

        <FormRow label="Room">
          <Select
            id="roomId"
            options={options!}
            ariaLabel="RoomId"
            optionsConfigForm={{
              valueAsNumber: true,
              onChange: () => trigger('roomId'),
            }}
          />
        </FormRow>

        <FormRow label="Address" error={errors.address?.message}>
          <TextArea
            id="address"
            rows={3}
            {...register('address', {
              required: REQUIRED_FIELD_ERROR,
              validate: {
                checkValidString: (v) => isValidString(v) || INVALID_FIELD,
              },
              onChange: () => trigger('address'),
            })}
          />
        </FormRow>

        <Form.Action>
          <FormBtn type="submit" name="submit" disabled={!isDirty || !isValid}>
            {
              // prettier-ignore
              isAdd
              ? 'Add'
              : 'Save'
            }
          </FormBtn>
          <FormBtn type="button" styled="secondary" onClick={onClose}>
            Close
          </FormBtn>
        </Form.Action>
      </Form>
    </FormProvider>
  );
};

export default UserForm;
