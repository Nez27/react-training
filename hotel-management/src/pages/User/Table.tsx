import { useSearchParams } from 'react-router-dom';
import { Dispatch, SetStateAction, useEffect, useState } from 'react';
import toast from 'react-hot-toast';

// Components
import { RiEditBoxFill } from 'react-icons/ri';
import { IoExit } from 'react-icons/io5';
import Menus from '../../components/Menus';
import Table from '../../components/Table';
import Direction from '../../commons/styles/Direction';
import Message from '../../components/Message';
import Search from '../../components/Search';
import SortBy from '../../components/SortBy';
import OrderBy from '../../components/OrderBy';

// Types
import { Nullable } from '../../types/common';
import { TUser } from '../../types/user';

// Hooks
import { useFetch } from '../../hooks/useFetch';

// Constants
import { STATUS_CODE } from '../../constants/responseStatus';
import { ORDERBY_OPTIONS, USER_PAGE } from '../../constants/variables';
import { CONFIRM_MESSAGE } from '../../constants/messages';

// Styled
import { StyledOperationTable } from './styled';
import Spinner from '../../commons/styles/Spinner';

// Services
import { updateRoomStatus } from '../../services/roomServices';
import { checkOutUser } from '../../services/userServices';

interface IUserRow {
  user: TUser;
  openFormDialog: () => void;
  setUser: Dispatch<SetStateAction<Nullable<TUser>>>;
  reload: boolean;
  setReload: Dispatch<SetStateAction<boolean>>;
}

const UserRow = ({
  user,
  openFormDialog,
  setUser,
  reload,
  setReload,
}: IUserRow) => {
  const handleEdit = (user: TUser) => {
    setUser(user);
    openFormDialog();
  };

  const handleCheckOut = async (user: TUser) => {
    if (confirm(CONFIRM_MESSAGE)) {
      const resUpdateStatus = await updateRoomStatus(user.roomId, false);
      const resCheckoutUser = await checkOutUser(user);

      if (
        resCheckoutUser?.statusCode === STATUS_CODE.OK &&
        resUpdateStatus?.statusCode === STATUS_CODE.OK
      ) {
        toast.success('Check out complete!');

        // Reload table
        setReload(!reload);
      }
    }
  };

  const { id, name, identifiedCode, phone, roomId } = user;

  return (
    <Table.Row>
      <div>{id}</div>
      <div>{name}</div>
      <div>{identifiedCode}</div>
      <div>{phone}</div>
      <div>{
        roomId 
          ? roomId 
          : 'None'
      }</div>

      <Menus.Menu>
        <Menus.Toggle id={id.toString()} />

        <Menus.List id={id.toString()}>
          <Menus.Button
            icon={<RiEditBoxFill />}
            onClick={() => handleEdit(user)}
          >
            Edit
          </Menus.Button>
          <Menus.Button icon={<IoExit />} onClick={() => handleCheckOut(user)}>
            Check out
          </Menus.Button>
        </Menus.List>
      </Menus.Menu>
    </Table.Row>
  );
};

interface IUserTable {
  reload: boolean;
  setReload: Dispatch<SetStateAction<boolean>>;
  openFormDialog: () => void;
  setUser?: Dispatch<SetStateAction<Nullable<TUser>>>;
}

const UserTable = ({
  reload,
  setReload,
  openFormDialog,
  setUser,
}: IUserTable) => {
  const [users, setUsers] = useState<TUser[]>([]);
  const [phoneSearch, setPhoneSearch] = useState('');
  const [searchParams] = useSearchParams();
  const sortByValue = searchParams.get('sortBy')
    ? searchParams.get('sortBy')!
    : '';
  const orderByValue = searchParams.get('orderBy')
    ? searchParams.get('orderBy')!
    : '';

  const { data, isPending, errorFetchMsg } = useFetch(
    'users',
    'phone',
    phoneSearch,
    sortByValue,
    orderByValue,
    reload
  );

  useEffect(() => {
    if (data) {
      setUsers(data);
    } else {
      setUsers([]);
    }

    if (errorFetchMsg) {
      console.error(errorFetchMsg);
    }
  }, [data, errorFetchMsg, setUsers]);

  return (
    <>
      <Direction>
        <StyledOperationTable>
          <OrderBy options={ORDERBY_OPTIONS} />

          <SortBy options={USER_PAGE.SORTBY_OPTIONS} />
          <Search
            setValueSearch={setPhoneSearch}
            setPlaceHolder="Search by phone..."
          />
        </StyledOperationTable>

        {isPending && <Spinner />}

        {users.length ? (
          <Menus>
            <Table columns="10% 30% 20% 20% 10% 5%">
              <Table.Header>
                <div>Id</div>
                <div>Name</div>
                <div>Identified Code</div>
                <div>Phone</div>
                <div>Room Id</div>
              </Table.Header>
              <Table.Body<TUser>
                data={users}
                render={(user: TUser) => (
                  <UserRow
                    user={user}
                    key={user.id}
                    reload={reload}
                    setReload={setReload}
                    openFormDialog={openFormDialog}
                    setUser={setUser!}
                  />
                )}
              />
            </Table>
          </Menus>
        ) : (
          !isPending && <Message>No data to show here!</Message>
        )}
      </Direction>
    </>
  );
};

export default UserTable;
