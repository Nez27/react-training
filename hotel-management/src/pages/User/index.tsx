import { useRef, useState } from 'react';

// Components
import Direction from '../../commons/styles/Direction';
import Button from '../../commons/styles/Button';
import UserTable from './Table';

// Styled
import { StyledUser, Title } from './styled';
import UserDialog from './Dialog';

const User = () => {
  const dialogRef = useRef<HTMLDialogElement>();
  const [reload, setReload] = useState(true);

  const openDialog = () => {
    dialogRef.current?.showModal();
  };

  const closeDialog = () => {
    dialogRef.current?.close();
  };

  return (
    <>
      <StyledUser>
        <Direction type="horizontal">
          <Title>List User</Title>
          <Button onClick={openDialog}>Add user</Button>
        </Direction>

        <UserTable reload={reload} />
      </StyledUser>

      <UserDialog
        onClose={closeDialog}
        ref={dialogRef}
        setReload={setReload}
        reload={reload}
      />
    </>
  );
};

export default User;
