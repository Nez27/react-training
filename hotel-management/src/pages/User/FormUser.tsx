import {
  Dispatch,
  SetStateAction,
  useCallback,
  useEffect,
  useState,
} from 'react';
import toast from 'react-hot-toast';

// Hooks
import { FormProvider, useForm } from 'react-hook-form';

// Styled
import Input from '../../commons/styles/Input.ts';

// Components
import Form from '../../components/Form/index.tsx';
import FormRow from '../../components/LabelControl/index.tsx';
import Select, { ISelectOptions } from '../../components/Select/index.tsx';

// Helpers
import {
  isEmptyObj,
  isValidName,
  isValidNumber,
  isValidPhoneNumber,
} from '../../helpers/validators.ts';

// Constants
import { ADD_SUCCESS, EDIT_SUCCESS } from '../../constants/messages.ts';
import {
  INVALID_FIELD,
  INVALID_PHONE,
  REQUIRED_FIELD_ERROR,
} from '../../constants/formValidateMessage.ts';
import { INIT_VALUE_USER_FORM } from '../../constants/variables.ts';

// Styled
import { FormBtn } from './styled.ts';

// Services
import { getAllRoom, updateRoomStatus } from '../../services/roomServices.ts';

// Types
import { Nullable } from '../../types/common.ts';
import { IUser } from '../../types/users.ts';
import { IRoom } from '../../types/rooms.ts';
import { createUser, updateUser } from '../../services/userServices.ts';

interface IUserFormProp {
  onCloseModal?: () => void;
  reload: boolean;
  setReload: Dispatch<SetStateAction<boolean>>;
  user?: Nullable<IUser>;
}

const UserForm = ({ onCloseModal, reload, setReload, user }: IUserFormProp) => {
  const formMethods = useForm<IUser>();
  const {
    register,
    handleSubmit,
    reset,
    formState: { errors, isDirty, isValid, isSubmitting },
    trigger,
  } = formMethods;
  const [rooms, setRooms] = useState<IRoom[]>([]);
  const [options, setOptions] = useState<ISelectOptions[]>();

  // Init value when edit form and load options
  useEffect(() => {
    const load = async () => {
      const options: ISelectOptions[] = [];
      const tempUser = !user 
        ? { roomId: 0 } 
        : { ...user };

      // Load and set default options room
      if (rooms.length > 0) {
        // Init first options
        options.push({
          label: '---Select---',
          value: '0',
        });

        rooms.forEach((item) => {
          if (!item.status || tempUser?.roomId === item.id)
            options.push({
              label: item.name!,
              value: item.id!.toString(),
            });
        });
      }

      if (options.length > 0) {
        setOptions(options);
      }

      if (!isEmptyObj(tempUser)) {
        // Init value
        reset(tempUser);
      } else {
        reset(INIT_VALUE_USER_FORM);
      }
    };

    load();
  }, [reset, user, rooms]);

  // Submit form
  const onSubmit = useCallback(
    async (newUser: IUser) => {
      try {
        if (!user) {
          // Add request
          const response = await createUser(newUser);

          if (response) {
            toast.success(ADD_SUCCESS);
          }

          // Update room status
          updateRoomStatus(newUser.roomId, true);
        } else {
          // Edit request
          const response = await updateUser(newUser);

          if (response) {
            toast.success(EDIT_SUCCESS);
          }

          // Update room status
          updateRoomStatus(user!.roomId, true, newUser.roomId);
        }
        // Reload table data
        setReload(!reload);
      } catch (error: unknown) {
        if (error instanceof Error) {
          toast.error(error.message);
        }
      }

      reset();
      onCloseModal!();
    },
    [onCloseModal, reload, reset, setReload, user]
  );

  // Load all rooms
  useEffect(() => {
    const loadRoom = async () => {
      const rooms = await getAllRoom();

      if (rooms) {
        setRooms(rooms);
      }
    };

    loadRoom();
  }, [onSubmit]);

  return (
    <FormProvider {...formMethods}>
      <Form onSubmit={handleSubmit(onSubmit)}>
        <Input type="hidden" id="id" {...register('id')} />
        <FormRow label="Full Name" error={errors?.name?.message}>
          <Input
            type="text"
            id="name"
            {...register('name', {
              required: REQUIRED_FIELD_ERROR,
              validate: {
                checkValidName: (value) => isValidName(value) || INVALID_FIELD,
              },
              onChange: () => trigger('name'),
            })}
          />
        </FormRow>

        <FormRow
          label="Identified Code"
          error={errors?.identifiedCode?.message}
        >
          <Input
            type="text"
            id="identifiedCode"
            {...register('identifiedCode', {
              required: REQUIRED_FIELD_ERROR,
              validate: {
                checkIdentifiedCode: (v) =>
                  isValidNumber(v.toString()) || INVALID_FIELD,
              },
              onChange: () => trigger('identifiedCode'),
            })}
          />
        </FormRow>

        <FormRow label="Phone" error={errors?.phone?.message}>
          <Input
            type="text"
            id="phone"
            {...register('phone', {
              required: REQUIRED_FIELD_ERROR,
              validate: {
                checkPhoneNum: (v) => isValidPhoneNumber(v) || INVALID_PHONE,
              },
              onChange: () => trigger('phone'),
            })}
          />
        </FormRow>

        <FormRow label="Room">
          {options && options.length > 1 ? (
            <Select
              id="roomId"
              options={options!}
              ariaLabel="RoomId"
              optionsConfigForm={{
                valueAsNumber: true,
                onChange: () => trigger('roomId'),
                validate: (v) => v !== 0,
              }}
            />
          ) : (
            <p>No room available!</p>
          )}
        </FormRow>

        <Form.Action>
          <FormBtn type="submit" name="submit" disabled={!isDirty || !isValid || isSubmitting}>
            {
              !user 
                ? 'Add' 
                : 'Save'
            }
          </FormBtn>
          <FormBtn type="button" styled="secondary" onClick={onCloseModal}>
            Close
          </FormBtn>
        </Form.Action>
      </Form>
    </FormProvider>
  );
};

export default UserForm;
