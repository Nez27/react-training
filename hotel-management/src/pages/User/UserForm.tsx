// Hooks
import { FormProvider, useForm } from 'react-hook-form';
import { useCreateUser } from '@hook/users/useCreateUser.ts';
import { useUpdateUser } from '@hook/users/useUpdateUser.ts';

// Styled
import Input from '@commonStyle/Input.ts';

// Components
import Form from '@component/Form/index.tsx';

// Helpers
import { isValidRegex } from '@helper/validators.ts';

import {
  INVALID_FIELD,
  INVALID_PHONE,
  REQUIRED_FIELD_ERROR,
} from '@constant/formValidateMessage.ts';
import { REGEX } from '@constant/commons.ts';

// Styled
import { FormBtn } from './styled.ts';

// Types
import { IUser } from '@type/users.ts';

interface IUserFormProp {
  onCloseModal?: () => void;
  user?: IUser;
}

const UserForm = ({ onCloseModal, user }: IUserFormProp) => {
  const { isCreating, createUser } = useCreateUser();
  const { isUpdating, updateUser } = useUpdateUser();
  const isLoading = isCreating || isUpdating;
  const {id: editId, ...editValues} = {...user};
  
  const formMethods = useForm<IUser>({
    defaultValues: editId 
      ? editValues 
      : {}
  });
  const {
    register,
    handleSubmit,
    reset,
    formState: { errors, isDirty, isValid },
    trigger,
  } = formMethods;

  // Submit form
  const onSubmit = 
    (newUser: IUser) => {
      if (!user) {
        // Add request
        createUser(newUser, {
          onSuccess: () => {
            reset();
            onCloseModal?.();
          },
        });
      } else {
        // Edit request
        newUser.id = editId!;
        updateUser(newUser, {
          onSuccess: () => {
            reset();
            onCloseModal?.();
          }
        });
      }
    };

  return (
    <FormProvider {...formMethods}>
      <Form onSubmit={handleSubmit(onSubmit)}>
        <Form.Row label="Full Name" error={errors?.name?.message}>
          <Input
            type="text"
            id="name"
            {...register('name', {
              required: REQUIRED_FIELD_ERROR,
              validate: {
                checkValidName: (value) =>
                  isValidRegex(new RegExp(REGEX.NAME), value) || INVALID_FIELD,
              },
              onChange: () => trigger('name'),
            })}
          />
        </Form.Row>

        <Form.Row label="Phone" error={errors?.phone?.message}>
          <Input
            type="text"
            id="phone"
            {...register('phone', {
              required: REQUIRED_FIELD_ERROR,
              validate: {
                checkPhoneNum: (v) =>
                  isValidRegex(new RegExp(REGEX.PHONE), v) || INVALID_PHONE,
              },
              onChange: () => trigger('phone'),
            })}
          />
        </Form.Row>

        <Form.Action>
          <FormBtn
            type="submit"
            name="submit"
            disabled={!isDirty || !isValid || isLoading}
          >
            {
              !user 
                ? 'Add' 
                : 'Save'
            }
          </FormBtn>
          <FormBtn type="button" variations="secondary" onClick={onCloseModal}>
            Close
          </FormBtn>
        </Form.Action>
      </Form>
    </FormProvider>
  );
};

export default UserForm;
