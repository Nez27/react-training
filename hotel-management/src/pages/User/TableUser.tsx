import { useSearchParams } from 'react-router-dom';
import { Dispatch, SetStateAction, useEffect, useState } from 'react';

// Components
import Menus from '../../components/Menus';
import Table from '../../components/Table';
import Direction from '../../commons/styles/Direction';
import Message from '../../components/Message';
import Search from '../../components/Search';
import SortBy from '../../components/SortBy';
import OrderBy from '../../components/OrderBy';

// Types
import { IUser } from '../../types/users';

// Hooks
import { useFetch } from '../../hooks/useFetch';

// Constants
import { ORDERBY_OPTIONS, USER_PAGE } from '../../constants/variables';

// Styled
import { StyledOperationTable } from './styled';
import Spinner from '../../commons/styles/Spinner';
import UserRow from './UserRow';





interface IUserTable {
  reload: boolean;
  setReload: Dispatch<SetStateAction<boolean>>;
}

const UserTable = ({ reload, setReload }: IUserTable) => {
  const [users, setUsers] = useState<IUser[]>([]);
  const [phoneSearch, setPhoneSearch] = useState('');
  const [searchParams] = useSearchParams();
  const sortByValue = searchParams.get('sortBy')
    ? searchParams.get('sortBy')!
    : '';
  const orderByValue = searchParams.get('orderBy')
    ? searchParams.get('orderBy')!
    : '';

  const { data, isPending, errorFetchMsg } = useFetch(
    'users',
    'phone',
    phoneSearch,
    sortByValue,
    orderByValue,
    reload
  );

  useEffect(() => {
    if (data) {
      setUsers(data);
    } else {
      setUsers([]);
    }

    if (errorFetchMsg) {
      console.error(errorFetchMsg);
    }
  }, [data, errorFetchMsg, setUsers]);

  return (
    <>
      <Direction>
        <StyledOperationTable>
          <OrderBy options={ORDERBY_OPTIONS} />

          <SortBy options={USER_PAGE.SORTBY_OPTIONS} />
          <Search
            setValueSearch={setPhoneSearch}
            setPlaceHolder="Search by phone..."
          />
        </StyledOperationTable>

        {isPending && <Spinner />}

        {users.length ? (
          <Menus>
            <Table columns="10% 30% 20% 20% 10% 5%">
              <Table.Header>
                <div>Id</div>
                <div>Name</div>
                <div>Identified Code</div>
                <div>Phone</div>
                <div>Room Id</div>
              </Table.Header>
              <Table.Body<IUser>
                data={users}
                render={(user: IUser) => (
                  <UserRow
                    user={user}
                    key={user.id}
                    reload={reload}
                    setReload={setReload}
                  />
                )}
              />
            </Table>
          </Menus>
        ) : (
          !isPending && <Message>No data to show here!</Message>
        )}
      </Direction>
    </>
  );
};

export default UserTable;
