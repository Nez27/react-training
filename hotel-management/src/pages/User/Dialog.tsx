import { forwardRef, useEffect } from 'react';

// Components
import Dialog from '../../components/Dialog';
import UserForm from './Form';

// Interfaces
import { IDialogProps } from '../../globals/interfaces';

// Types
import { TUser } from '../../globals/types';

const UserDialog = forwardRef((props, ref) => {
  const dialogRef = ref as React.MutableRefObject<
    HTMLDialogElement | undefined
  >;

  // prettier-ignore
  const { 
    onClose,
    setReload,
    reload,
    data,
    isAdd 
  } = props;

  useEffect(() => {
    if (dialogRef.current) {
      dialogRef.current.addEventListener('click', (e: MouseEvent) => {
        const dialogDimensions = dialogRef.current!.getBoundingClientRect();
        if (
          e.clientX < dialogDimensions.left ||
          e.clientX > dialogDimensions.right ||
          e.clientY < dialogDimensions.top ||
          e.clientY > dialogDimensions.bottom
        ) {
          dialogRef.current!.close();
        }
      });
    }
  }, [dialogRef]);

  return (
    <Dialog title={'Add user'} onClose={onClose} ref={dialogRef}>
      <UserForm
        onClose={onClose!}
        reload={reload!}
        setReload={setReload!}
        user={data}
        isAdd={isAdd!}
      />
    </Dialog>
  );
}) as React.FC<IDialogProps<TUser>>;

export default UserDialog;
