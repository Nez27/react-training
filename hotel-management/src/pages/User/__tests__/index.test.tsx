import renderer from 'react-test-renderer';
import User from '..';
import { BrowserRouter } from 'react-router-dom';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';

describe('User', () => {
  const queryClient = new QueryClient();

  const wrapper = renderer.create(
    <QueryClientProvider client={queryClient}>
      <BrowserRouter>
        <User />
      </BrowserRouter>
    </QueryClientProvider>
  );

  test('Should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
