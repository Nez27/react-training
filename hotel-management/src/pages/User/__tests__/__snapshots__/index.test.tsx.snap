// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Should render correctly 1`] = `
<main
  className="sc-lmgPJu fmDjvX"
>
  <div
    className="sc-cxNGUP cGfCbe"
    type="horizontal"
  >
    <h2
      className="sc-iJCSeZ jfWGKb"
    >
      List User
    </h2>
    <button
      className="sc-iemXqs gdUlEE"
      onClick={[Function]}
      variations="primary"
    >
      Add user
    </button>
  </div>
  <div
    className="sc-cxNGUP hDpxxT"
    type="vertical"
  >
    <div
      className="sc-giAruI kILZpe"
    >
      <div
        className="sc-dItHI exgtuU"
      >
        <button
          className="sc-bqGGcD entqcN"
          disabled={true}
          onClick={[Function]}
        >
          Ascending
        </button>
        <button
          className="sc-bqGGcD graLyf"
          disabled={false}
          onClick={[Function]}
        >
          Descending
        </button>
      </div>
      <select
        aria-label="Sort"
        className="sc-crzpnZ iXFhDe"
        disabled={false}
        onChange={[Function]}
        value=""
      >
        <option
          data-testid="select-option"
          value="id"
        >
          Sort by id
        </option>
        <option
          data-testid="select-option"
          value="name"
        >
          Sort by name
        </option>
        <option
          data-testid="select-option"
          value="phone"
        >
          Sort by phone
        </option>
      </select>
      <input
        className="sc-iqAbyq kOrhXJ"
        onChange={[Function]}
        placeholder="Search by phone..."
      />
    </div>
    <div
      className="sc-ezyZrH dFKszz"
    />
  </div>
</main>
`;
